internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_events.go

```diff
@@ -45,48 +43,56 @@ import (
+//                                       pos = 43                           e.writePos = 50
 //
-// Capacity of Events in this diagram is 18, Publisher published already 14 + 2 * 18 = 40 events, so it
-// already overwrote `e.stream` twice fully (e.gen = 2).
+// Capacity of Events in this diagram is 18, Publisher published already 50 events, so it
+// already overwrote `e.stream` twice fully.
 //
-// Consumer1 is trying to keep up with the publisher, it's on the same `gen`, and it has 14-7 = 7 events
-// to catch up.
+// Consumer1 is trying to keep up with the publisher, it has 14-7 = 7 events to catch up.
 //
-// Consumer2 is on `gen` 1, so it is reading events which were published before the Publisher did last
-// wraparound for `e.writePos` at `e.gen == 1`. Consumer 2 has a lot of events to catch up, but as it stays
+// Consumer2 is reading events published by Publisher before last wraparound, it has
+// 50-34 = 16 events to catch up. Consumer 2 has a lot of events to catch up, but as it stays
 // on track, it can still do that.
 //
-// Consumer3 is doing bad - it's on `gen` 1, but Publisher already overwrote this element while going on `gen` 2,
-// so Consumer3 is handling incorrect data, it should error out.
+// Consumer3 is doing bad: 50-27 = 23 > 18 (capacity), so its read position has already been
+// overwritten, it can't read consistent data,  soit should error out.
 //
-// Synchronization: at the moment single mutex protects `e.stream`, `e.writePos` and `e.gen`, consumers keep their
+// Synchronization: at the moment single mutex protects `e.stream`  and `e.writePos`, consumers keep their
 // position as local variable, so it doesn't require synchronization. If Consumer catches up with Publisher,
 // it sleeps on condition variable to be woken up by Publisher on next publish.
 type Events struct {
 	// stream is used as ring buffer of events
 	stream []runtime.Event
 
 	// writePos is the index in streams for the next write (publish)
-	writePos int
-
-	// gen tracks number of wraparounds in stream
-	gen int64
+	//
+	// writePos gets always incremented, real position in slice is (writePos % cap)
+	writePos int64
 
-	// cap is capacity of streams
+	// cap is a capacity of the stream
 	cap int
+	// gap is a safety gap between consumers and publishers
+	gap int
 
-	// mutext protects access to writePos, gen and stream
+	// mutext protects access to writePos and stream
 	mu sync.Mutex
 	c  *sync.Cond
 }
 
 // NewEvents initializes and returns the v1alpha1 runtime event stream.
-func NewEvents(cap int) *Events {
+//
+// Argument cap is a maximum event stream capacity (available event history).
+// Argument gap is a safety gap to separate consumer from the publisher.
+// Maximum available event history is (cap-gap).
+func NewEvents(cap, gap int) *Events {
 	e := &Events{
 		stream: make([]runtime.Event, cap),
 		cap:    cap,
+		gap:    gap,
+	}
+
+	if gap >= cap {
+		panic("NewEvents: gap >= cap")
```

Comment:
Do we want to panic here?

Reply:
I thought we should never get to this condition, as `Events` is internal and not configurable directly, but if `gap >= cap`, things will go completely nuts in this file, so I thought there should be some protection.

Reply:
:+1: